File: indexes/cpp/src/pq.cpp
#include "pq.hpp"

#include <algorithm>
#include <chrono>
#include <stdexcept>
#include <thread>

#include "distance.hpp"

namespace vro::pq {
namespace {

double seconds_since(std::chrono::steady_clock::time_point t0) {
    return std::chrono::duration<double>(std::chrono::steady_clock::now() - t0).count();
}

}  // namespace

Index build(Matrix& vectors, const Params& params, const BuildContext& ctx) {
    const long long m_in = params.get_int("m");
    const long long nbits = params.get_int("nbits");
    const std::string& metric = params.get_string("metric");
    const long long train_size = params.get_int("train_size");
    const long long iters = params.get_int("iters");
    if (nbits != 8) throw std::invalid_argument("pq: nbits must be 8, got " + std::to_string(nbits));
    if (metric != "ip" && metric != "l2")
        throw std::invalid_argument("pq: metric must be ip or l2, got " + metric);
    if (m_in <= 0 || vectors.dim % static_cast<std::size_t>(m_in) != 0)
        throw std::invalid_argument("pq: m must divide dim " + std::to_string(vectors.dim) +
                                    ", got " + std::to_string(m_in));
    if (train_size <= 0) throw std::invalid_argument("pq: train_size must be > 0");
    if (iters <= 0) throw std::invalid_argument("pq: iters must be > 0");

    Index idx;
    idx.vectors = &vectors;
    idx.n = vectors.rows;
    idx.dim = vectors.dim;
    idx.m = static_cast<std::size_t>(m_in);
    idx.dsub = idx.dim / idx.m;
    idx.metric = metric == "l2" ? Metric::kL2 : Metric::kDot;
    const std::size_t m = idx.m, dsub = idx.dsub, n = idx.n;
    const int threads = std::max(1, ctx.threads);

    // Train: one k-means per sub-vector position on the first train_n rows.
    auto t0 = std::chrono::steady_clock::now();
    const std::size_t train_n = std::min(n, static_cast<std::size_t>(train_size));
    idx.codebooks.resize(m * kCentroids * dsub);
    std::vector<float> sub(train_n * dsub);
    for (std::size_t j = 0; j < m; ++j) {
        for (std::size_t i = 0; i < train_n; ++i)
            std::copy_n(vectors.row(i) + j * dsub, dsub, sub.data() + i * dsub);
        KMeansOptions opt;
        opt.k = kCentroids;
        opt.iters = static_cast<int>(iters);
        opt.train_size = train_n;
        opt.seed = ctx.seed + j;
        opt.threads = threads;
        opt.metric = idx.metric;
        opt.normalize = false;
        std::vector<float> centers = kmeans(sub.data(), train_n, dsub, opt);
        std::copy(centers.begin(), centers.end(), idx.codebooks.begin() + j * kCentroids * dsub);
    }
    idx.times.train_s = seconds_since(t0);

    // Add: encode every row, rows split across threads.
    t0 = std::chrono::steady_clock::now();
    idx.codes.resize(n * m);
    auto encode = [&](std::size_t lo, std::size_t hi) {
        for (std::size_t i = lo; i < hi; ++i) {
            const float* x = vectors.row(i);
            for (std::size_t j = 0; j < m; ++j)
                idx.codes[i * m + j] = static_cast<std::uint8_t>(
                    nearest_center(x + j * dsub, idx.codebooks.data() + j * kCentroids * dsub,
                                   kCentroids, dsub, idx.metric));
        }
    };
    const std::size_t t_count = std::min<std::size_t>(static_cast<std::size_t>(threads), std::max<std::size_t>(n, 1));
    std::vector<std::thread> pool;
    const std::size_t chunk = (n + t_count - 1) / t_count;
    for (std::size_t t = 0; t < t_count; ++t) {
        std::size_t lo = t * chunk, hi = std::min(n, lo + chunk);
        if (lo < hi) pool.emplace_back(encode, lo, hi);
    }
    for (auto& th : pool) th.join();
    idx.times.add_s = seconds_since(t0);
    return idx;
}

SearchResult search(const Index& index, const float* query, std::size_t k, const Params& params) {
    const long long rerank_in = params.get_int("rerank");
    if (rerank_in < 0) throw std::invalid_argument("pq: rerank must be >= 0");
    const std::size_t rerank = static_cast<std::size_t>(rerank_in);
    const std::size_t m = index.m, dsub = index.dsub, n = index.n;
    const bool l2 = index.metric == Metric::kL2;

    // Table T (m, 256): higher is better in both modes.
    std::vector<float> table(m * kCentroids);
    for (std::size_t j = 0; j < m; ++j) {
        const float* qj = query + j * dsub;
        const float* cb = index.codebooks.data() + j * kCentroids * dsub;
        for (std::size_t c = 0; c < kCentroids; ++c)
            table[j * kCentroids + c] = l2 ? -l2_sq(qj, cb + c * dsub, dsub) : dot(qj, cb + c * dsub, dsub);
    }

    TopK top(rerank > 0 ? rerank : k);
    const float* t = table.data();
    const std::uint8_t* codes = index.codes.data();
    for (std::size_t i = 0; i < n; ++i) {
        const std::uint8_t* code = codes + i * m;
        float s = 0.0f;
        for (std::size_t j = 0; j < m; ++j) s += t[j * kCentroids + code[j]];
        if (s >= top.threshold()) top.push(static_cast<std::int64_t>(i), s);
    }

    SearchResult r;
    r.distance_computations = static_cast<std::int64_t>(n);
    if (rerank == 0) {
        top.result(r.ids, r.scores);
        return r;
    }
    std::vector<std::int64_t> cand;
    std::vector<float> approx;
    top.result(cand, approx);
    TopK exact(k);
    const Matrix& v = *index.vectors;
    for (std::int64_t id : cand) {
        if (id < 0) continue;
        const float* x = v.row(static_cast<std::size_t>(id));
        float s = l2 ? -l2_sq(query, x, index.dim) : dot(query, x, index.dim);
        exact.push(id, s);
        ++r.distance_computations;
    }
    exact.result(r.ids, r.scores);
    return r;
}

std::size_t index_bytes(const Index& index) {
    return index.codebooks.size() * sizeof(float) + index.codes.size();
}

std::map<std::string, double> extra(const Index& /*index*/) { return {}; }

}  // namespace vro::pq

File: indexes/python/pq.py
"""Product quantization index (CONTRACT 6.4).

Recall floor (CONTRACT 9, dev set): defaults, recall@10 >= 0.50 (metric=ip and metric=l2).

Build: split each vector into m sub-vectors of dim/m values. Train one codebook of 256
centroids per sub-vector with the shared k-means (seed + j, no normalization, assignment
by `metric`). Encode every row as m uint8 codes (the index of the best centroid).
Search: build a table T (m, 256) from the query, score row i as sum_j T[j, code[i, j]].
"""

import time
from concurrent.futures import ThreadPoolExecutor

import numpy as np

from . import distance, kmeans

BUILD_PARAMS: dict = {"m": 48, "nbits": 8, "metric": "ip", "train_size": 100000, "iters": 20}
SEARCH_PARAMS: dict = {"rerank": 0}

ENCODE_CHUNK = 65536  # rows encoded per step, to bound the (rows, 256) score matrix


def _check(params: dict, dim: int) -> tuple[int, str]:
    m, nbits, metric = int(params["m"]), params["nbits"], params["metric"]
    if nbits != 8:
        raise ValueError(f"pq: nbits={nbits} is not supported, only 8")
    if metric not in ("ip", "l2"):
        raise ValueError(f"pq: metric={metric!r} is not supported, want ip or l2")
    if m <= 0 or dim % m != 0:
        raise ValueError(f"pq: dim {dim} is not divisible by m={m}")
    return m, metric


def train_codebook(train: np.ndarray, iters: int, seed: int, metric: str) -> np.ndarray:
    """CONTRACT 6.4: the shared k-means with k=256, assignment by `metric`, no normalization."""
    return kmeans.kmeans(train, 256, iters=iters, seed=seed, train_size=len(train), metric=metric, normalize=False)


def encode(vectors: np.ndarray, codebooks: np.ndarray, metric: str) -> np.ndarray:
    """Return codes (N, m) uint8: best centroid per sub-vector under the metric."""
    n, dim = vectors.shape
    m, _, sub = codebooks.shape
    codes = np.empty((n, m), dtype=np.uint8)
    for start in range(0, n, ENCODE_CHUNK):
        block = vectors[start : start + ENCODE_CHUNK]
        for j in range(m):
            labels, _ = kmeans.assign(block[:, j * sub : (j + 1) * sub], codebooks[j], metric)
            codes[start : start + len(block), j] = labels
    return codes


def build(vectors: np.ndarray, params: dict, threads: int, seed: int) -> dict:
    n, dim = vectors.shape
    m, metric = _check(params, dim)
    sub = dim // m
    train_n = min(n, int(params["train_size"]))

    t0 = time.perf_counter()
    train = vectors[:train_n]
    codebooks = np.empty((m, 256, sub), dtype=np.float32)

    def train_one(j: int) -> None:
        sub_train = np.ascontiguousarray(train[:, j * sub : (j + 1) * sub], dtype=np.float32)
        codebooks[j] = train_codebook(sub_train, int(params["iters"]), seed + j, metric)

    # The m codebooks are independent (own seed, own slice), so the result does not depend
    # on the thread count. NumPy releases the GIL in its inner loops.
    with ThreadPoolExecutor(max_workers=max(1, int(threads))) as pool:
        list(pool.map(train_one, range(m)))
    train_s = time.perf_counter() - t0

    t0 = time.perf_counter()
    codes = encode(vectors, codebooks, metric)
    add_s = time.perf_counter() - t0

    return {
        "vectors": vectors,  # full vectors, only for rerank
        "codebooks": codebooks,
        "codebook_sq": np.einsum("jcs,jcs->jc", codebooks, codebooks),  # ||c||^2 for l2 tables
        "codes": codes,
        "offsets": np.arange(m, dtype=np.intp) * 256,
        "m": m,
        "metric": metric,
        "train_s": train_s,
        "add_s": add_s,
    }


def table(index: dict, query: np.ndarray) -> np.ndarray:
    """T (m, 256): ip -> q_j . c; l2 -> -||q_j - c||^2 (CONTRACT 6.4.1)."""
    cb = index["codebooks"]
    m, _, sub = cb.shape
    qs = query.reshape(m, sub)
    dots = np.einsum("jcs,js->jc", cb, qs)
    if index["metric"] == "ip":
        return dots
    q_sq = np.einsum("js,js->j", qs, qs)[:, None]
    return -(q_sq - 2 * dots + index["codebook_sq"])


def search(index: dict, query: np.ndarray, k: int, params: dict) -> tuple[np.ndarray, np.ndarray]:
    codes = index["codes"]
    n, m = codes.shape
    t = table(index, query)
    # s[i] = sum_j T[j, codes[i, j]]: flat index j*256 + code into the raveled table.
    s = np.take(t.ravel(), codes + index["offsets"]).sum(axis=1, dtype=np.float32)  # (N,)
    rerank = int(params.get("rerank", 0))
    if rerank <= 0:
        index["distance_computations"] = n
        return distance.top_k(s, k)
    cand, _ = distance.top_k(s, rerank)
    cand = cand[cand >= 0]
    x = index["vectors"][cand]
    if index["metric"] == "ip":
        exact = x @ query
    else:
        diff = x - query
        exact = -np.einsum("ij,ij->i", diff, diff)
    index["distance_computations"] = n + len(cand)
    return distance.top_k(exact.astype(np.float32), k, cand)


def index_bytes(index: dict) -> int:
    return int(index["codebooks"].size * 4 + index["codes"].size)
